### ISSUE TITLE
More parameter converters

### ISSUE THREAD

**Issue by TiS** [sensiolabs/SensioFrameworkExtraBundle#393]
I have written a bundle with two more parameter converters: one for service and one for ArrayObject. Maybe you'd be interested in integrating this in FrameworkExtraBundle?

https://packagist.org/packages/gendoria/param-converter-bundle

In my personal case, I find especially ServiceParamConverter useful, as I don't use Doctrine repositories but my own.

**Comment by jakzal** [comment]
Few months back I have created a similar bundle: https://packagist.org/packages/zalas/param-converter-bundle

**Comment by TiS** [comment]
Yes, I have seen it. Mine is a little bit more complex, as the parameters for the service call can be not only request parameters, but also other services and parameters from options array. But anyway, the idea of service parameter converter is IMHO good, no matter, whose will be incorporated :)

**Comment by Ener-Getick** [comment]
I don't think that's a good idea to inject services that way...
An example is when a service is deprecated, you don't know it and your tests don't cover the place where it is used. You won't have any warning but you will have an error when this service will be removed and you'll try to access your website.

**Comment by TiS** [comment]
@Ener-Getick Why is that so? I retrieve service using container ->get method. Same, as in controller. If the service is deprecated, then you either have a deprecation warning in logs (in both cases - fetch directly in controller or in annotation service) or don't have it at all and your application crashes - and then, again, it doesn't matter, if you fetch the service in controller or in annotation service. 

If your tests cover controller action (integration tests), then you will have correct error. If they do not, you will fail. If you test controller action without testing its annotations and you use them - this can fail in so many different ways, that I think this one is IMHO irrelevant.

Please, describe exactly, why it is dangerous (or rather - differently dangerous than getting service in controller).

**Comment by Ener-Getick** [comment]
Fetching services from the container is also in most cases a bad practice. And it's weird to pass the dependencies of a class when you call one of its method...

**Comment by TiS** [comment]
Well, this is the case of nearly ALL annotation based configuration. Event 'classic' Doctrine converter annotation uses some services which can be changed or removed. Not to mention the case, when you use parameter converter annotation with security annotation.

In my opinion, all annotation configured controllers **have to** be tested using integration tests (in which case, there's no problem with testing this specific converter). You can still use unit tests for your controllers, as parameter converter just injects some dependencies you can inject manually.

The whole purpose of this bundle lies in extending Symfony configuration with some useful annotations. Thus, if you want to use and test **ANY** of this bundle features, you **HAVE TO** test it with integration tests.

Also, please, take a look at my converter's 'supports' function:

```php
  public function supports(ParamConverter $configuration)
    {
        $options = $this->getOptions($configuration);
        return $this->container->has($options['service']) && !empty($options['method']);
    }
```

As you see, it makes proper checks, if the required service and method exists. It can also be extended to check for proper function arguments and argument types, if needed.

Although, I have to reluctanctly agree, that it is somewhat less secure to use, than other parameter converters. One has to wage the pros and cons. But I think it is good to give developer a choice, especially given the extreme usefullness of this converter (**it covers all cases, when one wants to use @security and other annotations, but does not use standard doctrine configuration as backend storage of entities**; this happens more often, that one may think).

In summary. I still think, that service parameter converter (and maybe ArrayObject parameter converter, other provided by my bundle) are very interesting and needed addition to this bundle. With fair warnings about the need of integration tests in documentation :).

**Comment by fabpot** [comment]
Closing as Symfony 3.3 and 3.4 cover almost everything here.
